lib/extract/netcdf/extract_netcdf_base.jinja

```diff
@@ -0,0 +1,464 @@
+    !! @param[in] name The name of the variable (string).
+    !! @param[in] value The value of the variable.
+    subroutine WriteArray{{dim}}d{{name}}(this, name, value)
+        use netcdf, only: nf90_put_var
+        implicit none
+        class(ExtractNetcdfBaseType), intent(inout), target :: this
+        character(*), intent(in) :: name
+        {{type}}, dimension({{DIMENSION}}), intent(in) :: value
+
+        integer :: retval
+
+        if (.not. is_enabled) return
+        retval = CheckError(nf90_put_var(this%ncid, this%var_id(this%next_var_index),  &
+                                         value({{DIMENSION}})))
+        call this%PSyDataBaseType%ProvideArray{{dim}}d{{name}}(name, value)
+
+    end subroutine WriteArray{{dim}}d{{name}}
+
+    ! -------------------------------------------------------------------------
+    !> This subroutine reads the values of a dl_esm_inf field (r2d_field).
+    !! It allocates a 2d-double precision array to store the read values
+    !! which is then returned to the caller. If the memory for the array can
+    !! not be allocated, the application will be stopped.
+    !! @param[inout] this The instance of the extract_PsyDataType.
+    !! @param[in] name The name of the variable (string).
+    !! @param[out] value An allocatable, unallocated 2d-double precision array
+    !!             which is allocated here and stores the values read.
+    subroutine ReadArray{{dim}}d{{name}}(this, name, value)
+        use netcdf
+        implicit none
+
+        class(ExtractNetcdfBaseType), intent(inout), target :: this
+        character(*), intent(in) :: name
+        {{type}}, dimension({{DIMENSION}}), allocatable, intent(out) :: value
+        integer :: retval, varid
+        integer :: dim_id
+        {# Create list of variables: "dimid1,dimid2,dimid3,dimid4" #}
+        {% set dim_sizes = "dim_size"~range(1,dim+1)|join(",dim_size") %}
+        integer :: {{dim_sizes}}
+        integer :: ierr
+        character(100) :: x
+
+        ! First query the dimensions of the original r2d_field from the
+        ! netcdf file
+        {% for i in range(1, dim+1) %}
+        retval = CheckError(nf90_inq_dimid(this%ncid, trim(name//"dim{{i}}"),  &
+                                           dim_id))
+        retval = CheckError(nf90_inquire_dimension(this%ncid, dim_id,  &
+                                                   len=dim_size{{i}}))
+        {% endfor %}
+
+        ! Allocate enough space to store the values to be read:
+        allocate(value({{dim_sizes}}), Stat=ierr)
+        if (ierr /= 0) then
+            print *,"Cannot allocate array for ", name, &
+                    " of size ", {{dim_sizes}}, " in ReadArray{{dim}}d{{name}}."
+            stop
+        endif
+        ! Initialise it with 0, so that an array comparison will work
+        ! even though e.g. boundary areas or so might not be set at all.
+        value = 0.0d0
```

Comment:
Is this hardwired double-precision constant OK here? Presumably the compiler will just do the Right Thing?

Reply:
Yes, the compiler should automatically convert this to the appropriate type.